Framework/Algorithms/src/PolarizationCorrections/PolarizationEfficienciesWildes.cpp

```diff
@@ -0,0 +1,407 @@
+    calculatePolarizerAndAnalyserEfficiencies(solveForP, solveForA);
+  }
+
+  progress.report(8, "Setting algorithm outputs");
+  setOutputs();
+}
+
+namespace {
+void setUnitAndDistributionToMatch(const MatrixWorkspace_sptr &wsToUpdate, const MatrixWorkspace_sptr &matchWs) {
+  wsToUpdate->setYUnit(matchWs->YUnit());
+  wsToUpdate->setDistribution(matchWs->isDistribution());
+}
+} // unnamed namespace
+
+void PolarizationEfficienciesWildes::calculateFlipperEfficienciesAndPhi() {
+  // Calculate the polarizing and analysing flipper efficiencies
+  const WorkspaceGroup_sptr nonMagWsGrp = getProperty(PropNames::INPUT_NON_MAG_WS);
+  const auto &flipperConfig = getPropertyValue(PropNames::FLIPPERS);
+  const auto &ws00 = workspaceForSpinState(nonMagWsGrp, flipperConfig, SpinStateValidator::ZERO_ZERO);
+  const auto &ws01 = workspaceForSpinState(nonMagWsGrp, flipperConfig, SpinStateValidator::ZERO_ONE);
+  const auto &ws10 = workspaceForSpinState(nonMagWsGrp, flipperConfig, SpinStateValidator::ONE_ZERO);
+  const auto &ws11 = workspaceForSpinState(nonMagWsGrp, flipperConfig, SpinStateValidator::ONE_ONE);
+
+  const auto numerator = ws00 - ws01 - ws10 + ws11;
+
+  const auto ws00Minus01 = ws00 - ws01;
+  const auto ws00Minus10 = ws00 - ws10;
+
+  m_wsFp = numerator / (2 * ws00Minus01);
+  m_wsFa = numerator / (2 * ws00Minus10);
+
+  // Calculate phi
+  m_wsPhi = (ws00Minus01 * ws00Minus10) / ((ws00 * ws11) - (ws01 * ws10));
+}
+
+MatrixWorkspace_sptr PolarizationEfficienciesWildes::calculateTPMOFromPhi(const WorkspaceGroup_sptr &magWsGrp) {
+  const auto &flipperConfig = getPropertyValue(PropNames::FLIPPERS);
+  const auto &ws00 = workspaceForSpinState(magWsGrp, flipperConfig, SpinStateValidator::ZERO_ZERO);
+  const auto &ws01 = workspaceForSpinState(magWsGrp, flipperConfig, SpinStateValidator::ZERO_ONE);
+  const auto &ws10 = workspaceForSpinState(magWsGrp, flipperConfig, SpinStateValidator::ONE_ZERO);
+  const auto &ws11 = workspaceForSpinState(magWsGrp, flipperConfig, SpinStateValidator::ONE_ONE);
+
+  // We use the flipper efficiency to multiply the mag ws counts, but the resulting workspace will have lost the Y unit
+  // and distribution information. We need to put these back otherwise the rest of the calculation fails when it tries
+  // to add and subtract workspaces with different Y units.
+  const auto twoFp = 2 * m_wsFp;
+  const auto twoFa = 2 * m_wsFa;
+
+  const auto twoFa00 = (1 - twoFa) * ws00;
+  setUnitAndDistributionToMatch(twoFa00, ws00);
+
+  const auto twoFa10 = (twoFa - 1) * ws10;
+  setUnitAndDistributionToMatch(twoFa10, ws10);
+
+  const auto twoFp00 = (1 - twoFp) * ws00;
+  setUnitAndDistributionToMatch(twoFp00, ws00);
+
+  const auto twoFp01 = (twoFp - 1) * ws01;
+  setUnitAndDistributionToMatch(twoFp01, ws01);
+
+  const auto numerator = twoFa00 + twoFa10 - ws01 + ws11;
+  const auto denominator = twoFp00 + twoFp01 - ws10 + ws11;
+  const auto tpmoSquared = m_wsPhi * (numerator / denominator);
```

Comment:
I thought about the same thing following the useful insight you got on performance from the work you did on that other PR. On balance I decided against it here, though, because looking more closely at the add/multiply etc. Mantid methods I felt that they do some useful validation and include some other logic that I wasn't keen to either duplicate or go without in this particular case. On real data the algorithm takes approximately 30 seconds to run, which is acceptable for the purpose of calibration (it should be run a maximum of once per experiment, although as I understand it probably more like once per cycle, and likely less than that as currently POLREF are intending to continue using their own implementation, which runs at the same sort of speed). If performance ever becomes more of a consideration then we can come back to it at that stage.

It might be worth flagging your findings on performance of Mantid workspace arithmetic to the Core team/more generally, though. I think for long-term maintainability it would be better for a more central solution to be found to improve workspace arithmetic performance and then have that made available to all algorithms.